client/my-sites/plans-v2/upsell.tsx

```diff
@@ -1,41 +1,187 @@
  */
-import React from 'react';
+import React, { useCallback } from 'react';
 import { useSelector } from 'react-redux';
 import page from 'page';
+import { useTranslate } from 'i18n-calypso';
 
 /**
  * Internal dependencies
  */
-import { slugToSelectorProduct, durationToString } from './utils';
+import { ProductIcon } from '@automattic/components';
+import Gridicon from 'components/gridicon';
+import FormattedHeader from 'components/formatted-header';
+import JetpackProductCard from 'components/jetpack/card/jetpack-product-card';
+import Main from 'components/main';
+import { addItem, addItems } from 'lib/cart/actions';
+import { jetpackProductItem } from 'lib/cart-values/cart-items';
+import { preventWidows } from 'lib/formatting';
+import { getCurrentUserCurrencyCode } from 'state/current-user/selectors';
+import { isProductsListFetching, getAvailableProductsList } from 'state/products-list/selectors';
 import { getSelectedSiteSlug } from 'state/ui/selectors';
+import {
+	durationToString,
+	durationToText,
+	getProductPrices,
+	getProductUpsell,
+	slugToSelectorProduct,
+} from './utils';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
 
 /**
  * Type dependencies
  */
-import type { UpsellPageProps } from './types';
+import type { SelectorProduct, UpsellPageProps } from './types';
+
+// There are several things that could go wrong depending on the URL
+// - The product slug could not exist.
+// - The product slug could be of a product that does not have another product to
+// upsell to – Jetpack Anti-Spam is an example.
+// - The product slug could have an upsell product but what if the user already has that
+// product?
+// - What if the user already owns the product associated with the product slug?
+
+interface Props {
+	currencyCode: string;
+	mainProduct: SelectorProduct;
+	selectedSiteSlug: string;
+	upsellProduct: SelectorProduct;
+}
+
+const UpsellComponent = ( {
+	currencyCode,
+	mainProduct,
+	selectedSiteSlug,
+	upsellProduct,
+}: Props ) => {
+	const translate = useTranslate();
+
+	const availableProducts = useSelector( ( state ) => getAvailableProductsList( state ) );
+
+	// Upsell prices should come from the API with the upsell products
+	const upsellProductPrices = getProductPrices( upsellProduct, availableProducts );
+	const savedAmount = 100;
+
+	const { shortName: mainProductName, productSlug } = mainProduct;
+	const { shortName: upsellProductName, productSlug: upsellProductSlug } = upsellProduct;
+
+	const goToCheckout = useCallback( () => {
+		page( `/checkout/${ selectedSiteSlug }` );
+	}, [ selectedSiteSlug ] );
+
+	const onPurchaseBothProducts = useCallback( () => {
+		addItems( [ jetpackProductItem( productSlug ), jetpackProductItem( upsellProductSlug ) ] );
+		goToCheckout();
+	}, [ goToCheckout, productSlug, upsellProductSlug ] );
+
+	const onPurchaseSingleProduct = useCallback( () => {
+		addItem( jetpackProductItem( productSlug ) );
+		goToCheckout();
+	}, [ goToCheckout, productSlug ] );
```

Comment:
Curious why the code uses `useCallback` here? Does any of these need memoization?

Reply:
Not memoization but on every render those functions are going to be recreated making children components re-render even if the rest of the props haven't changed. I'll say that I'm used to do it. On higher components in the components tree could be troublesome because by definition those will have many children.

On the other hand, I don't think having those could hurt. What do you think?